pennylane/data/attributes/dictionary.py

```diff
@@ -0,0 +1,83 @@
+# Copyright 2018-2023 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Contains an AttributeType that allows for heterogenous dictionaries
+of Dataset attributes."""
+
+
+from collections.abc import Mapping, MutableMapping
+from typing import Generic, Optional, TypeVar, Union
+import typing
+from pennylane.data.base.attribute import AttributeType
+from pennylane.data.base.mapper import MapperMixin
+from pennylane.data.base.typing_util import T, ZarrAny, ZarrGroup
+
+
+class DatasetDict(
+    Generic[T],
+    AttributeType[ZarrGroup, typing.Mapping[str, T], Optional[typing.Mapping[str, T]]],
+    MutableMapping,
+    MapperMixin,
+):
+    """Provides a dict-like collection for Dataset attribute types."""
+
+    Self = TypeVar("Self", bound="DatasetDict")
+
+    type_id = "dict"
+
+    def __post_init__(self, value: Optional[typing.Mapping[str, T]], info):
+        if value:
+            self.update(value)
```

Comment:
Value will always be passed here, since this method is called by `__init__`. Since it's not intended for direct use there's no need for default arguments.

Reply:
default value is now `{}`, which simplifies the typing a bit.